src/server/session.ts

```diff
@@ -531,26 +535,35 @@ namespace ts.server {
             defaultDefinition :
             defaultProject.getLanguageService().getSourceMapper().tryGetSourcePosition(defaultDefinition!));
 
-        // Track which projects we have already searched so that we don't repeat searches.
-        // We store the project key, rather than the project, because that's what `loadAncestorProjectTree` wants.
-        // (For that same reason, we don't use `resultsMap` for this check.)
-        const searchedProjects = new Set<string>();
+        // The keys of resultsMap allow us to check which projects have already been searched, but we also
+        // maintain a set of strings because that's what `loadAncestorProjectTree` wants.
+        const searchedProjectKeys = new Set<string>();
 
         onCancellation:
         while (queue.length) {
             while (queue.length) {
                 if (cancellationToken.isCancellationRequested()) break onCancellation;
 
+                let skipCount = 0;
+                for (; skipCount < queue.length && resultsMap.has(queue[skipCount].project); skipCount++);

```

Comment:
Maybe not worth it, but I wonder how often you end up with a run of projects. Might even be able to even avoid the Set lookup if you keep track of a `lastProject`.

Reply:
Why is it `resultsMap` here instead of `searchedProjectKeys`?

Reply:
The items in the queues are locations and not projects - that's why there are so many.  And they get enqueued before the corresponding projects are done processing, so they wouldn't be filtered.

`resultsMap` uses `Project` directly, rather than a project key (i.e. string) as in `searchedProjectKeys`.  The strings are only maintained for `loadAncestorProjects`.  There should be a comment to that effect above the declaration.